### PR TITLE
Implement the WAGED rebalancer with the limited functionality.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

#402 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The rebalancer will support the basic rebalance logic.
Note this implementation does not contain the logic to support delayed rebalance and user defined preference list.

### Tests

- [x] The following tests are written for this issue:

Add unit test to cover the main workflow of the WAGED rebalancer.

- [x] The following is the result of the "mvn test" command on the appropriate module:

[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 898, Failures: 1, Errors: 0, Skipped: 1

Re-run:
[INFO] Results:
[INFO] 
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

https://github.com/apache/helix/wiki/Weight-aware-Globally-Evenly-distributed-Rebalancer

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

**Comment by jiajunwang** [comment]
Test is not complete yet.
Also, I will refine the format before marking this PR ready for review. Please focus on the logic if you are reviewing the draft. Thanks.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableNode.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
  * This class represents a possible allocation of the replication.
  * Note that any usage updates to the AssignableNode are not thread safe.
  */
-public class AssignableNode {
+public class AssignableNode implements Comparable<AssignableNode> {
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableNode.java]: Not really, I think what you mentioned is AssignableReplica.

File: helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/constraints/MockRebalanceAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * A mock up rebalance algorithm for unit test.
+ */
+public class MockRebalanceAlgorithm implements RebalanceAlgorithm {
+  Map<String, ResourceAssignment> _resultHistory = null;
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/constraints/MockRebalanceAlgorithm.java]: Both fine for the tests. But an empty map would be better. Will change.

**Comment by i3wangyi** [comment]
Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary

**Comment by jiajunwang** [comment]
> Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary

This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.

**Comment by i3wangyi** [comment]
> > Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary
> 
> This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.

It's strange for the unit test, you can simply use mockito.mock and mock any behavior without a real working class. A real working class is painful when future developers found the unit test is broken and he needs to understand what's going on inside the unit test helper. I would try avoiding such complicate coupled pattern in unit test as much as possible. 

It makes sense if you want to have such a class for integration test

**Comment by jiajunwang** [comment]
> > > Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary
> > 
> > 
> > This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.
> 
> It's strange for the unit test, you can simply use mockito.mock and mock any behavior without a real working class. A real working class is painful when future developers found the unit test is broken and he needs to understand what's going on inside the unit test helper. I would try avoiding such complicate coupled pattern in unit test as much as possible.
> 
> It makes sense if you want to have such a class for integration test

Did you mean if the test breaks, people won't need to understand the mockito class and the mocked methods definition for debugging? I don't think so. Actually, since the mock classes are not returning static maps or lists, we will need to keep the logic. So even they are mockito classes, I will need to "implement" the method using the same code. Given this, making them mockito classes is not a good option from my point of view.

**Comment by i3wangyi** [comment]
> > > > Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary
> > > 
> > > 
> > > This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.
> > 
> > 
> > It's strange for the unit test, you can simply use mockito.mock and mock any behavior without a real working class. A real working class is painful when future developers found the unit test is broken and he needs to understand what's going on inside the unit test helper. I would try avoiding such complicate coupled pattern in unit test as much as possible.
> > It makes sense if you want to have such a class for integration test
> 
> Did you mean if the test breaks, people won't need to understand the mockito class and the mocked methods definition for debugging? I don't think so. Actually, since the mock classes are not returning static maps or lists, we will need to keep the logic. So even they are mockito classes, I will need to "implement" the method using the same code. Given this, making them mockito classes is not a good option from my point of view.

During any refactor works, the unit test could fail. If you're testing method A, it depends on dependency class B, do you want to mock object and control all behaviors of B or use a concrete 
B instance? If the test of method A fails, do you know is it because of the wrong implementations of B or your method has flaws? Adding a custom mock will certainly increase the complexity and break the "unit test" concept which means isolated test, regardless of dependencies.

An example:
```
public void testMethodA() {
   Service service = new Servcie(new MockB()) // MockB is a custom mock class you may or may not understand how it is implemented
   boolean result = service.doSomething();
}

public void testMethodA() {
   B mock = Mockito.mock(B.class);
   when(mock.runSomehing()).thenDoSomething(); // since B is a mockito mock, its behavior is totally under control by this test metho without understanding how B internally runs
   Service service = new Service(mock);
   boolean result = service.doSomething();
}
```

**Comment by jiajunwang** [comment]
> > > > > Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary
> > > > 
> > > > 
> > > > This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.
> > > 
> > > 
> > > It's strange for the unit test, you can simply use mockito.mock and mock any behavior without a real working class. A real working class is painful when future developers found the unit test is broken and he needs to understand what's going on inside the unit test helper. I would try avoiding such complicate coupled pattern in unit test as much as possible.
> > > It makes sense if you want to have such a class for integration test
> > 
> > 
> > Did you mean if the test breaks, people won't need to understand the mockito class and the mocked methods definition for debugging? I don't think so. Actually, since the mock classes are not returning static maps or lists, we will need to keep the logic. So even they are mockito classes, I will need to "implement" the method using the same code. Given this, making them mockito classes is not a good option from my point of view.
> 
> During any refactor works, the unit test could fail. If you're testing method A, it depends on dependency class B, do you want to mock object and control all behaviors of B or use a concrete
> B instance? If the test of method A fails, do you know is it because of the wrong implementations of B or your method has flaws? Adding a custom mock will certainly increase the complexity and break the "unit test" concept which means isolated test, regardless of dependencies.
> 
> An example:
> 
> ```
> public void testMethodA() {
>    Service service = new Servcie(new MockB()) // MockB is a custom mock class you may or may not understand how it is implemented
>    boolean result = service.doSomething();
> }
> 
> public void testMethodA() {
>    B mock = Mockito.mock(B.class);
>    when(mock.runSomehing()).thenDoSomething(); // since B is a mockito mock, its behavior is totally under control by this test metho without understanding how B internally runs
>    Service service = new Service(mock);
>    boolean result = service.doSomething();
> }
> ```

In the example you mentioned, mock class or mockito class look exactly the same to me. Because the code in service.doSomething() and the code in when(mock.runSomehing()).thenDoSomething() will be exactly the same. In the 2nd case, you are just wrapping the code with when(...).thenDoSomething(). If the mockito thenDoSomething() method was done in the wrong way, the dev who is doing the refactor need to check mockito class the same, right? It is not magic.

High-levelly speaking, a dev needs to understand test logic. It is not an option. And the mock class here is considered to be part of the test case. Note the mock class files were in the test path. There is no way to get confused with a real class for production.
Moreover, anyone who can understand the mockito class will be able to understand the mock class.

So I don't think what you said make sense.

**Comment by i3wangyi** [comment]
> > > > > > Will you wait for my algorithm implementation part to be checked in? I don't think creating a mock class is necessary
> > > > > 
> > > > > 
> > > > > This mock class is for unit test. Even after your code is ready, I will keep it for the test purpose. Integration test will be done in a different style. We will include real components and turn on a ZK server for that.
> > > > 
> > > > 
> > > > It's strange for the unit test, you can simply use mockito.mock and mock any behavior without a real working class. A real working class is painful when future developers found the unit test is broken and he needs to understand what's going on inside the unit test helper. I would try avoiding such complicate coupled pattern in unit test as much as possible.
> > > > It makes sense if you want to have such a class for integration test
> > > 
> > > 
> > > Did you mean if the test breaks, people won't need to understand the mockito class and the mocked methods definition for debugging? I don't think so. Actually, since the mock classes are not returning static maps or lists, we will need to keep the logic. So even they are mockito classes, I will need to "implement" the method using the same code. Given this, making them mockito classes is not a good option from my point of view.
> > 
> > 
> > During any refactor works, the unit test could fail. If you're testing method A, it depends on dependency class B, do you want to mock object and control all behaviors of B or use a concrete
> > B instance? If the test of method A fails, do you know is it because of the wrong implementations of B or your method has flaws? Adding a custom mock will certainly increase the complexity and break the "unit test" concept which means isolated test, regardless of dependencies.
> > An example:
> > ```
> > public void testMethodA() {
> >    Service service = new Servcie(new MockB()) // MockB is a custom mock class you may or may not understand how it is implemented
> >    boolean result = service.doSomething();
> > }
> > 
> > public void testMethodA() {
> >    B mock = Mockito.mock(B.class);
> >    when(mock.runSomehing()).thenDoSomething(); // since B is a mockito mock, its behavior is totally under control by this test metho without understanding how B internally runs
> >    Service service = new Service(mock);
> >    boolean result = service.doSomething();
> > }
> > ```
> 
> In the example you mentioned, mock class or mockito class look exactly the same to me. Because the code in service.doSomething() and the code in when(mock.runSomehing()).thenDoSomething() will be exactly the same. In the 2nd case, you are just wrapping the code with when(...).thenDoSomething(). If the mockito thenDoSomething() method was done in the wrong way, the dev who is doing the refactor need to check mockito method the same, right? Mockito is not magic.
> 
> High-levelly speaking, a dev needs to understand test logic. It is not an option. And the mock class here is considered to be part of the test case. Note the mock class files were in the test path. There is no way to get confused with a real class for production.
> Moreover, anyone who can understand the mockito class will be able to understand the mock class.
> 
> So I don't think what you said make sense.

```
public void testMethodA() {
   Service service = new Servcie(new MockB()) // MockB is a custom mock class you may or may not understand how it is implemented
   boolean result = service.doSomething();
}

public void testMethodA() {
   B mock = Mockito.mock(B.class);
   when(mock.runSomehing()).thenDoSomething(); // since B is a mockito mock, its behavior is totally under control by this test method without understanding how B internally runs
   Service service = new Service(mock);
   boolean result = service.doSomething();
}
```

In the above first case, if the output result is not expected. It's probably for either 2 reasons:
1. The MockB doesn't do it right, it's possible
2. The MethodA is implemented wrong. 
Then the developer who writes the test has to dig out what's the real reason.

However, in the 2nd case, the developer knows exactly what the mockito mock class will do (because he knows just by checking the line above). The 1st possible reason in case 1 is eliminated.

In addition,  with mockito mock, you can easily let the mock class behavior unusually. E.g,
`when(mock.runSomehing()).thenThrowException();`
With a simple custom mock, probably you have to write the method again to test cases when the exception is thrown.

The reason behind it is natural with a trained eyes. Sometimes when the implementation of original class is very complex or not designed for easy unit testing, it will be difficult using mockito, I agree. In that case, a custom mock can take place.

**Comment by jiajunwang** [comment]
This PR is ready to be merged, approved by @narendly

**Pull request merged**